chain/cardano/client/client.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	logger := log.WithFields(log.Fields{
+		"chain":   cfg.Chain,
+		"rpc":     url,
+		"network": network,
+	})
+
+	return &Client{
+		Asset:               cfgI,
+		Url:                 url,
+		Network:             network,
+		Logger:              logger,
+		BlockfrostProjectId: cfg.Auth2.LoadOrBlank(),
+		HttpClient:          http.DefaultClient,
+	}, nil
+}
+
+func (client *Client) FetchProtocolParameters(ctx context.Context) (types.ProtocolParameters, error) {
+	var protocolParameters types.ProtocolParameters
+	err := client.Request(ctx, GET, "/epochs/latest/parameters", nil, &protocolParameters)
+
+	return protocolParameters, err
+}
+
+func (client *Client) FetchUtxos(ctx context.Context, address xc.Address, contract xc.ContractAddress) ([]types.Utxo, error) {
+	path := fmt.Sprintf("/addresses/%s/utxos/%s?order=desc", string(address), string(contract))
+	var response []types.Utxo
+	err := client.Request(ctx, GET, path, nil, &response)
+	return response, err
+}
+
+// 1. Sort utxos by amount descending
+// 2. Get minimum utxo set that adds up to the target amount
+func GetMinUtxoSet(utxos []types.Utxo, targetAmount xc.AmountBlockchain, contract xc.ContractAddress) []types.Utxo {
+	slices.SortFunc(utxos, func(lhs types.Utxo, rhs types.Utxo) int {
+		amountL := lhs.GetAssetAmount(contract)
+		amountR := rhs.GetAssetAmount(contract)
+
+		return amountR.Cmp(&amountL)
+	})
+
+	utxoSet := make([]types.Utxo, 0)
+	balance := xc.NewAmountBlockchainFromUint64(0)
+	for _, utxo := range utxos {
+		if balance.Cmp(&targetAmount) >= 0 {
+			break
+		}
+
+		amount := utxo.GetAssetAmount(contract)
+		balance = balance.Add(&amount)
+		utxoSet = append(utxoSet, utxo)
+	}
+	return utxoSet
+}
+
+// FetchTransferInput returns tx input for a Cardano transfer
+func (client *Client) FetchTransferInput(ctx context.Context, args xcbuilder.TransferArgs) (xc.TxInput, error) {
+	decimals, err := client.FetchDecimals(ctx, Lovelace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch decimals: %w", err)
+	}
+
+	contract, ok := args.GetContract()
+	if !ok {
+		contract = Lovelace
+	}
+
+	utxos, err := client.FetchUtxos(ctx, args.GetFrom(), contract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch utxos: %w", err)
+	}
+
+	// Include the fee in target amount
+	cfg := client.Asset.GetChain()
+	feeLimit := cfg.FeeLimit
+	xfeeLimit := feeLimit.ToBlockchain(int32(decimals))
+	targetAmount := args.GetAmount()
+	targetAmount.Add(&xfeeLimit)
+
+	utxos = GetMinUtxoSet(utxos, targetAmount, contract)
+
+	var latestBlock types.Block
+	err = client.Request(ctx, GET, "/blocks/latest", nil, &latestBlock)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch block info: %w", err)
+	}
+
+	protocolParams, err := client.FetchProtocolParameters(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch protocol parameters: %w", err)
+	}
+
+	baseFee := protocolParams.FixedFee
+	feePerByte := protocolParams.FeePerByte
```

Comment:
probably should instantiate a txBuilder + Tx to figure out the transaction size.